Added LZ77 complexity algorithm

HuwCheston opened the pull request:
This PR adds the complexity measure based on the LZ77 compression algorithm from [Cheston, Schlichting, Cross, & Harrison (2024)](https://royalsocietypublishing.org/doi/10.1098/rsos.240920). Input is a discrete 1D `numpy` array with any dtype, which is converted internally into a string representation based on the number of unique elements. Output is the length of the compressed string after applying LZ77.

pmcharrison commented on musmart/algorithm/complexity.py:
```diff
@@ -0,0 +1,127 @@
+
+    Parameters
+    ----------
+    sequence : np.ndarray
```
The types in the docstring should match the type hints in the source code. Or you could just delete the former.

HuwCheston replied:
Resolved in https://github.com/music-computing/amads/pull/14/commits/3cdcd889f4c77a7900fc1b8d69414faae1a4429c

MarkGotham commented:
Nudge @HuwCheston to keep this on the radar.

Are you up for finishing this off? 

Looks like a small and uncontroversial PR edit ... and you have first refusal on implementing as it's your PR ;)

HuwCheston commented on musmart/algorithm/complexity.py:
```diff
@@ -0,0 +1,87 @@
+        >>> lz77(np.array(["a"]))
+        1
+
+        >>> lz77(np.array(["a", "b", "a"]))
```
Resolved in https://github.com/music-computing/amads/pull/14/commits/3cdcd889f4c77a7900fc1b8d69414faae1a4429c

HuwCheston commented:
Ok, added a few new commits that should allow any sequence of immutable objects to be passed in. Also added a decode method. Tested and works with e.g. strings `"abababab"`, lists of strings, floats, etc. Think that covers the use cases we'll probably want.

pmcharrison reviewed (approved):
Great job!
